### ISSUE TITLE
Modify validation for Phone Numbers

### ISSUE THREAD

**Issue by Telos-sa** [usnistgov/OSCAL#2041]
### User Story

Currently the listed type for metadata>parties>telephone-numbers>number is a string, which has the regular expression pattern "\S(.*\S)?":
<img width="554" alt="Screenshot 2024-09-13 at 5 29 58 PM" src="https://github.com/user-attachments/assets/e40930ee-0fa1-479f-bf85-8165abf62494">

When validating using the oscal-cli, this field instead uses the regex pattern "^[0-9]{3}[0-9]{1,12}$" for validation, and throws warnings for varying (fake) phone numbers:
<img width="953" alt="Screenshot 2024-09-13 at 5 36 51 PM" src="https://github.com/user-attachments/assets/9ffa50a7-bffa-44bf-87cb-4fb39133cc7b">

### Goals

Validate using the string regex pattern which is outlined in the schema
Or
Create a new datatype for telephone-numbers>number that lists the regex pattern that is actually being used for validation
<img width="298" alt="Screenshot 2024-09-13 at 5 44 06 PM" src="https://github.com/user-attachments/assets/7718f719-d9ad-4a65-beb5-a60ee3855d75">

### Acceptance Criteria

- [X] All [OSCAL website](https://pages.nist.gov/OSCAL) and readme documentation affected by the changes in this issue have been updated. Changes to the OSCAL website can be made in the docs/content directory of your branch.
- [ ] A Pull Request (PR) is submitted that fully addresses the goals of this User Story. This issue is referenced in the PR.
- [ ] The CI-CD build process runs without any reported errors on the PR. This can be confirmed by reviewing that all checks have passed in the PR.

(For reviewers: [The wiki has guidance](https://github.com/usnistgov/OSCAL/wiki/Issue-Review) on code review and overall issue review for completeness.)

**Comment by iMichaela** [comment]
@Telos-sa - can you please provide the version of the `oscal-cli` you are using. I am assuming it is a FedRAMP constraint. While a better implementation is desired, we would need to see if other international numbers are supported.

**Comment by iMichaela** [comment]
How will extensions be noted when the regex pattern "^[0-9]{3}[0-9]{1,12}$" is used? Also numbers with prefixes marked with () ?

**Comment by Telos-sa** [comment]
@iMichaela This is version 2.0.2 of the base OSCAL validator (no FedRAMP constraints were used). I downloaded the pre-built version from https://repo1.maven.org/maven2/dev/metaschema/oscal/oscal-cli-enhanced/2.0.2/ as outlined on https://github.com/metaschema-framework/oscal-cli. If this is an incorrect oscal-cli please let us know where we can go to install the correct.

**Comment by iMichaela** [comment]
NIST `oscal-cli` latest release is v1.0.3. The version you point out to was generated not by NIST so we have no control over the OSCAL metaschema definitions (models) used to generate that version. If this is the one provided by FedRAMP and generated under `metaschema-framework` repo, please raise those concerns with the developers. **There is no NIST oscal-cli v2.0.2**. This is the version FedRAMP demo-ed last week. Please raise those issues with FedRAMP because the OSCAL version (when constraints are not applied) should be the official OSCAL release v1.1.2 or earlier. 
Can you please provide the version of OSCAL displayed by the tool? Thank you.

**Comment by Telos-sa** [comment]
Apologies for the version confusion, I have installed v1.0.3 from https://github.com/usnistgov/oscal-cli:
<img width="571" alt="Screenshot 2024-09-16 at 3 38 05 PM" src="https://github.com/user-attachments/assets/c47a42ea-da0b-45d6-be0d-93bcc8604e4a">
The same warnings are still present in v1.0.3
<img width="1159" alt="Screenshot 2024-09-16 at 3 24 34 PM" src="https://github.com/user-attachments/assets/704168e0-05e5-4658-915c-a96d62d379ac">

It makes sense that telephone-numbers>number would have a different validation regex pattern than strings, but that is not outlined in the v1.1.2 OSCAL metaschema.

telephone-numbers section under parties:
```
"telephone-numbers":{
    "type":"array",
    "minItems":1,
    "items":{
        "$ref":"#field_oscal-metadata_telephone-number"
    }
}
```

Assembly for field_oscal-metadata_telephone-number:
```
"oscal-ssp-oscal-metadata:telephone-number":{
    "title":"Telephone Number",
    "description":"A telephone service number as defined by ITU-T E.164.",
    "$id":"#field_oscal-metadata_telephone-number",
    "type":"object",
    "properties":{
        "type":{
            "title":"type flag",
            "description":"Indicates the type of phone number.",
            "anyOf":[
                {
                    "$ref":"#/definitions/StringDatatype"
                },
                {
                    "enum":[
                        "home",
                        "office",
                        "mobile"
                    ]
                }
            ]
        },
        "number":{
            "$ref":"#/definitions/StringDatatype"
        }
    },
    "required":[
        "number"
    ],
    "additionalProperties":false
}
```

StringDataType definition:
```
"StringDatatype":{
    "description":"A non-empty string with leading and trailing whitespace disallowed. Whitespace is: U+9, U+10, U+32 or [ \n\t]+",
    "type":"string",
    "pattern":"^\\S(.*\\S)?$"
}
```
The regex pattern "^[0-9]{3}[0-9]{1,12}$" is used by the oscal-cli for validation, but is not outlined anywhere in the metaschema.

**Comment by wendellpiez** [comment]
Hi, this is to confirm that the OSCAL metaschemas (and schemas) should not have any special validation for the lexical form of a `telephone-number`, so any such warning would presumably be coming from an OSCAL application or external validator (such as oscal-cli) enforcing its own rules (presumably in addition to OSCAL rules).

If the telephone numbers being flagged are correct, the test is wrong - perhaps because of a different scope or applicability such as US vs other. Please take this up with the developers as something that requires an adjustment or switch?

**Comment by iMichaela** [comment]
Since many numbers can be noted with extensions that need to be documented with the phone number, or other, a with a + sign in front for international numbers , or have area code in parentheses,  the warning (to me) is meant to raise human's attention to ensure the inserted number is correct, when non-numeric  values are used.  Tere is also the SIP protocol (https://en.wikipedia.org/wiki/SIP_URI_scheme) with valid format like sip:1-999-123-4567@voip-provider.example.net
Is the suggestion here to remove the regex="^[0-9]{3}[0-9]{1,12}$" constraint all together and not use it as a semi-validation mechanism for numeric-only phone numbers?

**Comment by GaryGapinski** [comment]
Ah, "telephone numbers". They can be and are anything you want them to be. The word _numbers_ is inadequate. Perhaps even _telephone_.

TL;DR: The constraint mentioned — `\S(.*\S)?` — is sufficient to encompass almost all telephone numbers of a length greater than 0 and is admirably catholic. It cannot express an absent value (the initial `\S` demands an initial non-space character). However, this constraint does not appear in the most recent ([v1.1.1](https://github.com/usnistgov/OSCAL/releases/tag/v1.1.2)) XML and JSON Schemas. The schema documentation mentions that `telephone-number` is "_A telephone service number as defined by ITU-T E.164._".

Without locale-specific presentation/formatting, even numeric telephone "numbers" look bad if subjected to artifical and unwarranted constraints.

Some references:
- Normative: [ITU-T E.164](https://www.itu.int/rec/T-REC-E.164/en). Some folks do not conform but happily make and receive calls anyway.
- Informative: https://en.wikipedia.org/wiki/E.164.
- Normative: [ITU-T E.123](https://www.itu.int/rec/T-REC-E.123). To borrow a phrase from Mark Twain, E.123 is commonly flung down and danced upon. ¹
- Informative: https://en.wikipedia.org/wiki/E.123.
- Normative: [RFC2806](https://datatracker.ietf.org/doc/html/rfc2806).
- Normative: [RFC3261](https://datatracker.ietf.org/doc/html/rfc3261). No such thing as all-numeric.
- Informative: https://en.wikipedia.org/wiki/SIP_URI_scheme.
- The above are, btw, over 20 years old. They lack currency and exhibit flaws and inconsistencies.
- Informative: https://en.wikipedia.org/wiki/National_conventions_for_writing_telephone_numbers.
- Normative: [RFC5724](https://www.rfc-editor.org/rfc/rfc5724).

Examples (mostly not E.123) (I frequently find demarcation of treasured "number" syllables via periods, hyphens, spaces, parentheses):

-  Deutsche [Bundesnachrichtendienst contact page](https://www.bnd.bund.de/EN/Contact/contact_node.html).
- [UK SIS (MI6) contact page](https://www.sis.gov.uk/contact-us.html).
- [CISA contact page](https://www.cisa.gov/about/contact-us).
- [GSA contact page](https://www.gsa.gov/about-us/contact-us?topnav=about-us).
- [Telos contact page](https://www.telos.com/contact-us/).

¹ [Fenimore Cooper's Literary Offences #7](https://www.gutenberg.org/files/3172/3172-h/3172-h.htm)

**Comment by wendellpiez** [comment]
Excellent point about validation against `\S(.*\S)?` - while noting that [metaschema-xslt](https://github.com/usnistgov/metaschema-xslt) (recently updated) has gotten some attention to datatypes - so cross-checking against its schemas (distinct from oscal-cli schemas or behavior) is also always warranted.

To the end of stabilizing testing and making cross-checks easier, if anyone is interested in schema field testing (either in theory or in practice), let me know or take a look at https://github.com/usnistgov/oscal-xproc3/tree/main/schema-field-tests

Fantastic info @GaryGapinski - thanks for the links.

**Comment by wendellpiez** [comment]
Noting in passing that the current `metaschema-xslt` also has a prototype implementation of constraints checking of OSCAL XML under XSLT. It is 95% feature-complete and testable, regex-matching being one of the things it is supposed to be able to do.

**Comment by iMichaela** [comment]
> Noting in passing that the current `metaschema-xslt` also has a prototype implementation of constraints checking under XSLT. It is 95% feature-complete and testable, regex-matching being one of the things it is supposed to be able to do. While it is untested outside the lab, trying OSCAL metaschema-based validation could be a solid next step.

@wendellpiez - can you please elaborate how are the warnings from oscal-cli handled in the metaschema-xslt ? In which branch do you have these latest updates?

**Comment by wendellpiez** [comment]
@iMichaela recently merged into the main branch of the in metaschema-xslt repository (called `develop`) is an experimental application to generate XSLTs that perform validation of metaschema rules including the constraints.

The "Inspector XSLT" generator works the same way as the schema generators do - you call it with your metaschema (from a command line or under CI/CD), and it generates an output, in this case not a schema but an XSLT 'schema emulator' that you can use to validate your OSCAL XML.

I'm happy to help anyone interested in testing. So far no one (tmk) has tried it except myself and one volunteer assistant, and on a testing metaschema, not OSCAL.

However, I would also caution that I built this originally as a proof of concept, and subsequently to work up my skills in unit testing - but it has no maintenance model and no 'team' to help (no users, only one dev). This makes me hesitant to offer it as a solution (even if it could be a 'next step' as I mentioned).

Indeed, another much simpler idea along the same lines is simply to cross-test the rules using Schematron - here is an example (testing port-range rules (an XSpec file for unit testing the Schematron implementation): https://github.com/wendellpiez/oscal-xproc3/blob/ssp-portrange-constraints-May2024/schema-field-tests/reference-sets/ssp-model/port-ranges-test.xspec - this would be much easier if you can work an XSpec.

**Comment by GaryGapinski** [comment]
[libphonenumber](https://github.com/google/libphonenumber) may be of interest to those with a need to manipulate phone numbers.

**Comment by aj-stein-gsa** [comment]
> Without locale-specific presentation/formatting, even numeric telephone "numbers" look bad if subjected to artifical and unwarranted constraints.
> 
> Some references:
> 
> * Normative: [ITU-T E.164](https://www.itu.int/rec/T-REC-E.164/en). Some folks do not conform but happily make and receive calls anyway.
> * Informative: https://en.wikipedia.org/wiki/E.164.
> * Normative: [ITU-T E.123](https://www.itu.int/rec/T-REC-E.123). To borrow a phrase from Mark Twain, E.123 is commonly flung down and danced upon. ¹
> * Informative: https://en.wikipedia.org/wiki/E.123.
> * Normative: [RFC2806](https://datatracker.ietf.org/doc/html/rfc2806).
> * Normative: [RFC3261](https://datatracker.ietf.org/doc/html/rfc3261). No such thing as all-numeric.
> * Informative: https://en.wikipedia.org/wiki/SIP_URI_scheme.
> * The above are, btw, over 20 years old. They lack currency and exhibit flaws and inconsistencies.
> * Informative: https://en.wikipedia.org/wiki/National_conventions_for_writing_telephone_numbers.
> * Normative: [RFC5724](https://www.rfc-editor.org/rfc/rfc5724).

I would agree with Gary and there is an excellent list of informative and normative references here. I can see a couple of solutions, and I can propose a PR for them.

Short-term solutions (backwards-compatible):

1. Relax the (recently identified) [`oscal-metadata-telephone-number-regex` constraint](https://github.com/usnistgov/OSCAL/blob/1eb7d81f8bc6b62dbdde7c3abbc2789a1c54914f/src/metaschema/oscal_metadata_metaschema.xml#L1036-L1041).
2. Add a `system` flag to the the telephone-number field for people to upgrade model versions (backwards compatible if not set) and allow them to, for example, pick `@system="http://ietf.org/rfc/rfc5724"` and then then implementers can have a telephone-number with that system to encode a phone number as a constraint URI `tel:+123456789` in those cases. We can add others according to desire.

Medium to long-term solutions (breaking backwards-compatibility):

3. Switch the core data type and make it URI, not string, and presume RFC5724 and document that into the model requirement. Many digital systems, including almost ubiquitous smart phones, support the `tel:` URI and it seems like that is the better play. I will not PR this approach, but I believe it is a better long-term approach that is not immediately actionable.
